Redesign monitored variables for hp_ilo sensor

bjw-s opened the pull request:
Allow generating specific sensors without the need for template sensors

## Description:
Major redesign on the hp_ilo component allowing generating specific sensors without the need for template sensors.

This means a small (but breaking) change for existing users, due to the `monitored_variables` configuration entry now requiring a sensor name and sensor type instead of being an plain list

I have added an example including a screenshot to the home-assistant.github.io PR.

**Related issue (if applicable):** n/a

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#2608

## Example entry for `configuration.yaml` (if applicable):
```yaml
sensor:
  - platform: hp_ilo
    host: IP_ADDRESS or HOSTNAME
    username: USERNAME
    password: PASSWORD
    monitored_variables:
      - name: SENSOR NAME
        sensor_type: SENSOR TYPE
```
## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

bjw-s committed:
Redesign monitored variables

Allow generating specific sensors without the need for template sensors

homeassistant commented:
Hi @Juggels, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#7534).

Once you do that we will be able to review and accept this pull request.

Thanks!

bjw-s committed:
Import 3rd party library inside update method

balloob commented on homeassistant/components/sensor/hp_ilo.py:
```diff
@@ -115,20 +135,25 @@ def device_state_attributes(self):
 
     def update(self):
         """Get the latest data from HP ILO and updates the states."""
+        import jsonpath_rw
```
Please don't include a new dependency for extracting data. We use Jinja2 templates for this.

bjw-s committed:
Remove jsonpath_rw dependency

Juggels commented:
@balloob I've done some changes, no longer requiring the external dependency. Also updated the docs in home-assistant/home-assistant#2608

balloob commented on homeassistant/components/sensor/hp_ilo.py:
```diff
@@ -84,15 +96,30 @@ def setup_platform(hass, config, add_devices, discovery_info=None):
         self.hp_ilo_data = hp_ilo_data
 
+        if sensor_value_template is not None:
+            # Make sure the template always returns valid json
```
We should not use Jinja2 templates to generate JSON that we then will load again. Jinja2 should be used to just generate the value.

Example template could be `{{ value.something[0].name }}`

balloob reviewed (changes requested):
We should not do any magic with incoming data.

Fetch data from device -> Run it through template -> Return it from a property.

Interpreting it as JSON is not allowed. Please remove.

bjw-s committed:
Do not interfere with value_template or ilo_data output

Do not interfere with value_template or ilo_data output, this is now the responsibility of the user and should be handled in `configuration.yaml`

Fix UnusedImportStatement

Fix newline after function docstring

Juggels commented:
Processes requested changes in latest commit

balloob commented on homeassistant/components/sensor/hp_ilo.py:
```diff
@@ -121,23 +143,15 @@ def update(self):
-                self._state = ilo_data
-        elif isinstance(ilo_data, (int, float)):
+        # the sensor state, otherwise store the data in the sensor attributes
+        if isinstance(ilo_data, (str, bytes, int, float)):
```
Can you remove this logic and always assign it to the state. I know that it's pre-existing logic but it shouldn't have been merged

balloob reviewed (changes requested):
Looks good. 1 minor change left.

bjw-s committed:
Always output results to state

Juggels commented:
Processes requested changes in latest commit

balloob commented:
Thanks! 🐬
